### PR TITLE
Rework epander / experiment / application logic

### PR DESCRIPTION
**Pull request by douglasjacobsen**
This merge is a major redesign of the logic to render, store, and expand experiments and the variables.

Fundamentally, this change creates an experiment set object which houses all of the experiments and the classes for each of them. This also pushes the variables into the application instances, and changes the expander class to just house the expansion logic instead of all of the variables as well.

### CODE REVIEW

**Comment by douglasjacobsen** [comment]
@rfbgo Sorry for how big this is... :)

**Comment by douglasjacobsen** [comment]
One thing I feel iffy about is `application._variables`. I can't decide if it should be `_variables` or `variables`. If you have any thoughts, feel free to pick all the nits about it!

**Review by rfbgo — APPROVED** [review]
Honestly, a really hard review with so much of the code moving files and so many API changes.. but I definitely agree with the spirit! 

Added 2 nits which you can ignore if you don't find them relevant

File: lib/ramble/ramble/experiment_set.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            tty.debug('Defining n_ranks in %s' %
+                      self.experiment_namespace)
+            variables[self.ranks_key] = n_ranks
+        elif not n_nodes:
+            variables[self.nodes_key] = 1
+
+        if not n_threads:
+            variables[self.threads_key] = 1
+
+    def _ingest_experiments(self):
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/experiment_set.py:198]: Yes, thanks for the reminder. I'll go through and update the comments in the PR in general.

**Review by rfbgo — APPROVED** [review]
LGTM!

**Pull request merged**